Fixed linked libs for win32

novist opened the pull request:
Closes #948
Closes #843

tux3 commented:
This PR and commit includes quite a lot of changes that are unrelated to the title and commit message. 
Please try to avoid that, now I have to try to carefully make sense of every unexplained change you made.

ghost commented:
A lot? Lets see...
- Fixed warning on windows [here](https://github.com/novist/qTox/commit/4c4b0f023009b688848972a07f79717712426b13#diff-dc7eafdfcdfc16aed754a99d1d3d7e8fL24)
- opencv2 includes modified in order to make it actually compile on windows because of how opencv installs itself. Even if stuff is disabled and not included in package main header still includes them. Its good mojo to include only whats needed anyway.
- install_libfilteraudio.sh fixed to work on windows. That should be covered by main msg i think.
- Removed unnecessarily linked libs because there are no such libs when building on win32. Those deps are statically linked into other dependencies and thats why it also failed to build on win32.

Yeah i indeed sneaked in that win32 warning fix. If it bad to discard very unimportant stuff as unimportant then guilty ;)

dubslow commented:
I don't like those modifications to install_libfilteraudio.sh; not providing an install location, I.e. using the default, shouldn't change the file extension... id rather see a dedicated option/argument to change the extension, or maybe modify windows/bootstrap.sh to rename the output

tux3 commented on qtox.pro:
```diff
@@ -74,9 +74,9 @@ contains(JENKINS,YES) {
-    LIBS += -L$$PWD/libs/lib -lopencv_core248 -lopencv_highgui248 -lopencv_imgproc248 -lOpenAL32 -lopus
-    LIBS += -lopengl32 -lole32 -loleaut32 -luuid -lvfw32 -ljpeg -ltiff -lpng -ljasper -lIlmImf -lHalf -lws2_32 -lz
+    LIBS += -L$$PWD/libs/lib -ltoxav -ltoxcore -ltoxencryptsave -ltoxdns -lsodium -lvpx -lpthread
+    LIBS += -L$$PWD/libs/lib -lopencv_core249 -lopencv_highgui249 -lopencv_imgproc249 -lOpenAL32 -lopus
```
This line will break the automated windows builds, that's not much of a problem, but it needs to be addressed before merging.

tux3 replied:
It's not a problem with the cross-compiler, just that many Linux distributions are still packaging OpenCV 2.4.8. An option named "QTOX_CROSSCOMPILING" would be misleading. "USE_OPENCV_248" maybe, but I don't think it's needed in the first place.
It just means that I'll have to build OpenCV 2.4.9 DLLs for the Jenkins server, and that's not a problem at all.

tux3 replied:
I'm pretty sure that " -ljpeg -ltiff -lpng -ljasper -lIlmImf -lHalf" are not actually needed, they are historical remains from when I tried to build OpenCV statically. And I should be able to update the build server to OpenCV 249

I'll make some tests and see what I can remove and what I can update.

tux3 commented:
The fix to src/platform/timer_win.cpp should probably go in a separate commit, I'd gladly merge that asap.

Same for the #include changes in src/video/camera*.h, they have nothing to do with linked libs, but it's an improvement :)

Thanks.

ghost commented:
@tux3 okie ill split them up ;)

@dubslow why do you prefer having install_libfilteraudio.sh broken on windows while it could be working?

EDIT: I see your point @dubslow. Will fix it ^_^

Commit:
Fixed warning in timer_win.cpp

Commit:
Including only required opencv headers. Workaround for opencv bug where partial builds would lack certain headers but opencv.hpp includes them anyway resulting in a build error.

Commit:
Updated linked libs for win32:
  * Removed libs that were not used
  * Reordered few libs to make mingw happy
  * Updated version of opencv libs to 249

Commit:
Added windows/bootstrap.sh which downloads and compiles all qtox dependencies
Fixed install_libfilteraudio.sh to work on windows
bootstrap.bat now invokes sh bootstrap.sh (for convenience)
Updated INSTALL.md to reflect simplified bootstrapping procedure

ghost commented:
@tux3 if there is any special handling for buildserver needed let me know what it is. Though i guess it would be perfect if buildserver could be accommodated to this build config so we dont have to use different one.

tux3 commented:
@novist I'm currently building OpenCV 2.4.9 on the build server, as soon as that's done I'll make some builds and test them.

tux3 commented:
Alright, the 32bit OpenCV 2.4.9 seems to build fine, but the -ws2_32 in the LIBS needs to come back or the build fails.

ghost commented:
Actually its [there](https://github.com/novist/qTox/commit/46cc81b4825b2d5b66cb0a0e76f8ec24d9ee7fde#diff-6a074c693ab3fcda2a48f60a4ef75bd5R79). Could you tell me where it should go to make build happy? I would test it on win32 too.

tux3 commented:
@novist Oh my bad, I had only seen the first two added lines, not the last unchanged one...
Alright then, the build server should be ready, this part is ok.

@dubslow  are you happy with the changes to the bootstrap scripts ? If yes, then let's merge it now, otherwise I can cherry-pick the 3 others now and keep the PR open for the bootstrap changes.

dubslow commented on windows/bootstrap.sh:
```diff
@@ -0,0 +1,109 @@
+
+## toxcore
+if [ ! -f "libtoxcore-win32-i686.zip" ]; then
+   wget --no-check-certificate http://jenkins.libtoxcore.so/job/libtoxcore-win32-i686/lastSuccessfulBuild/artifact/libtoxcore-win32-i686.zip
```
Not checking certificates is something that should be done with care, https is still useful, despite the flaws in the certificate system. (I assume it's related to running a Unix utility on Windows that SSL/TLS becomes Difficult™?)

dubslow replied:
That's basically what I expected. For now, I guess we'll just "oh well" this.
